Core/Spells: Paladin auras + gift of the naaru + penance

matanshukry opened the pull request:
**Changes proposed:**

-  Paladins auras: now only 1 can be active
-  Gift of the naaru now corerctly heals for 20% of the caster max health over 5 seconds
-  Gift of the naaru now also works for monk

**Target branch(es):** 3.3.5/master

- [ ] 3.3.5
- [x] master

**Issues addressed:**

Closes #26125
Closes #22468
Closes #21227

**Tests performed:**

- [x] It builds
- [x] Tested in-game

**Known issues and TODO list:**

matanshukry committed:
Correct paladin auras + gift of the naaru

mdX7 commented on src/server/scripts/Spells/spell_paladin.cpp:
```diff
@@ -184,29 +216,29 @@ class spell_pal_blessing_of_faith : public SpellScript
 
     void HandleDummy(SpellEffIndex /*effIndex*/)
     {
-        if (Unit* unitTarget = GetHitUnit())
```
indentation, also unrelated change i guess

matanshukry replied:
yup, not sure why it even changed; don't remember touching that. Thanks

mdX7 commented on src/server/scripts/Spells/spell_generic.cpp:
```diff
@@ -1322,28 +1329,7 @@ class spell_gen_gift_of_naaru : public AuraScript
-                break;
-        }
-
+        float heal = 0.2f * GetCaster()->GetMaxHealth();
```
we can get that 0.2f by reading basepoints of EFFECT1 / 100:
```
Effect 1: Id 3 (SPELL_EFFECT_DUMMY)
Difficulty: Id 0 (DIFFICULTY_NONE)
BasePoints = 20.00
Targets (0, 0) (NO_TARGET, NO_TARGET)
```

matanshukry committed:
Using effect1 value rather than hardcode + reverted indent

mdX7 commented on src/server/scripts/Spells/spell_generic.cpp:
```diff
@@ -1329,9 +1329,13 @@ class spell_gen_gift_of_naaru : public AuraScript
+        if (SpellEffectInfo const* eff1 = GetSpellInfo()->GetEffect(EFFECT_1))
+        {
+            float healPct = eff1->CalcValue() / 100.0f;
+            float heal = 0.2f * GetCaster()->GetMaxHealth();
```
you forgot to replace 0.2f with healPct here :D

matanshukry replied:
lol just saw that too xd

mdX7 commented on src/server/scripts/Spells/spell_generic.cpp:
```diff
@@ -1322,30 +1329,13 @@ class spell_gen_gift_of_naaru : public AuraScript
+            float healPct = eff1->CalcValue() / 100.0f;
+            float heal = 0.2f * GetCaster()->GetMaxHealth();
+            int32 healTick = std::floor(heal / aurEff->GetTotalTicks());
+            amount += int32(std::max(healTick, 0));
```
we may also remove that std::max here i guess?

matanshukry replied:
why?

mdX7 replied:
healTick will always be >= 0

matanshukry replied:
Unless `eff->CalcValue()` is negative (changed in the future) ?

mdX7 replied:
that'd require us to fix the spell either way (again) in that scenario, either it'll heal exactly 0 or it will deal damage (if < 0) so we might actually drop std::max for some minimal performance improvement

matanshukry replied:
cool, will do

matanshukry committed:
- Penance damage - remove caster as target
- gift_of_the_naaru - use calculated healPct value
- channel target selection - Don't log message if target is null due to script handler

matanshukry committed:
Removed max

Shauren committed:
Merge branch 'master' into issue/spells

Shauren commented:
Auras: a9fa007f83f99ccf38bc60bb1a75f671db092781
Gift of the Naaru: d2b70397ab98fe8a24a5880b757a50e3736787d3
Penance: not pushed - it looks like ChannelObjects implementnation is not 100% correct in Spell class
